docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,455 @@
+
+Note that the HelmReleaseProxy controller contains a static configuration and is only responsible for ensuring there is a Helm release matching the HelmReleaseProxy. The HelmReleaseProxy gets created, updated, and deleted by the HelmChartProxy controller.
+
+#### Considerations about the CRD design
+
+The scope of the cluster values accessible using Go templating must be defined along with use cases. An option would be to expose the Builtin struct type used in ClusterClass or construct a similar type. Additionally, while resources like the control plane and cluster have a 1:1 correlation with each selected cluster, resources like MachineDeployments or Machines could have any number and can’t be selected in a similar way. As a result, additional work is needed to decide if there’s a use case for including information in those types.
+
+It would also be useful to add `clusterv1.Conditions` to the HelmChartProxy and HelmReleaseProxy status fields to surface additional information about the status of the Helm release in a way that is consistent with the design of existing Cluster API CRDs.
+
+### Example: Installing Calico CNI
+
+Developers and Cluster operators often want to install a CNI on their workload clusters. HelmChartProxy controller can be used to install the Calico CNI on each selected workload cluster. In this example, the value configuration can be leveraged to initialize the `installation.calicoNetwork.ipPools` array by creating an element for every pod CIDR in the Cluster.
+
+```yaml
+apiVersion: addons.cluster.x-k8s.io/v1beta1
+kind: HelmChartProxy
+metadata:
+  name: calico-cni
+spec:
+  clusterSelector:
+    key: calicoCNI
+    value: enabled
+  releaseName: calico
+  repoURL: https://projectcalico.docs.tigera.io/charts
+  chartName: tigera-operator
+  values: |
+    installation:
+      cni:
+        type: Calico
+        ipam:
+          type: HostLocal
+      calicoNetwork:
+        bgp: Disabled
+        mtu: 1350
+        ipPools:{{range $i, $cidr := .Cluster.Spec.ClusterNetwork.Pods.CIDRBlocks }}
+        - cidr: {{ $cidr }}
+          encapsulation: None
+          natOutgoing: Enabled
+          nodeSelector: all(){{end}}
+```
+
+## Cluster API enhancements
+
+Despite the fact that most of the heavy lifting for add-on management is delegated to external ClusterAddonProvider components, over time we would like to provide Cluster API’s users an integrated and seamless user experience for add-on orchestration.
+
+### clusterctl
+
+ClusterAddonProvider is a new component to be installed in a Management Cluster, and in order to make this operation simpler, we can extend clusterctl in order to manage ClusterAddonProvider implementations in clusterctl init, clusterctl upgrade, etc.
+
+Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.
+
```

Comment:
Q: Are there any security implications / considerations that are worth mentioning?

Essentially we are introducing a component like tiller from helm v2 which was dropped because of security issues. I guess with the goals of an addon controller we basically have to run it in "god mode" (regarding the workload clusters) and it's probably fine as not everyone has access to the management cluster.

Anyway, just curious if there are any considerations worth mentioning (e.g. maybe something like auditing could be useful to provide visibility into what was done by the addon provider  in the future)

Reply:
I'm not too familiar with security in K8s. I believe @fabriziopandini added this section so I'll tag him. In terms of "god mode" I'm not too clear on why/if the add-on controller would be more insecure than the CRS controller if it the idea is to select clusters to install some arbitrary YAML.

Reply:
Not sure if grasp how the security concern relates to the paragraph above on clusterctl, but I'm ok in documenting that this  component is going to act on the workload clusters using the admin kubeconfig existing in the management cluster (but TBH this is the same of what any other component in CAPI is doing).

Reply:
I just added the comment here as the security section we have in other proposals not present. If we think there's nothing worth mentioning feel free to resolve.